chromaui/chromatic-e2e
# Picture elements are captured, including art direction

Issue: #AP-3783

## What Changed


We now properly snapshot `<picture>` elements.

## Background
A `<picture>` element can have multiple <source> child elements, but has to have exactly one `<img>` element (the img is used as a fallback in case none of the sources are supported or match the media queries/pixel densities provided).

## Our solution

Similar to @tevanoff's [img srcset fix](https://github.com/chromaui/chromatic-e2e/pull/40), we'll see if any of the `<picture>`'s `<source>` or `<img>` tags' srcset or src properties match an an item in our asset map. If so, we'll blow away all the `<source>`s of the `<picture>` tag and modify the existing `<img>` tag with the src set to the archived resource.

We take care not to modify _other_ attributes of the `<img>` tag, (like `class`), as we don't want to affect the original styling of the picture.


## How to test



We only need to manually test Cypress for mobile screens:
1. Create a new (or use an existing) Cypress Chromatic project
2. Create a page-under-test that renders a `<picture>` element with multiple `<source>`s that have differing image URLs -- one for desktop, one for mobile ([see here](https://github.com/chromaui/chromatic-e2e/blob/bfed37f1b1b54b027a413f8e063ff7b0b2afd9db/test-server/fixtures/asset-paths/picture.html) for example HTML, but better if you write your own)
3. Create a Cypress test that tests this page in mobile-sized view ([see here](https://github.com/chromaui/chromatic-e2e/blob/bfed37f1b1b54b027a413f8e063ff7b0b2afd9db/packages/cypress/tests/cypress/e2e/archiving-assets.cy.ts#L72-L76) for example)
4. Create a Cypress test that tests this page in desk-top sized view (can just omit the viewport options)
5. Run Cypress and Chromatic
4. In the small-test Chromatic snapshot, verify that the small-sized image is captured
4. In the large-test Chromatic snapshot, verify that the large-sized image is captured

The automated tests cover the rest:
* Unit tests assure the nuts-and-bolts of picture-source-finding work
* UI Tests assure that we 1) capture `<picture>` elements', well, pictures, as well as get the right one in the right circumstance
